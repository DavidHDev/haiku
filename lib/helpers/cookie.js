export const parseToDataType = (value, isItRetry = false) => {
  try {
    return value === 'undefined' ? undefined : JSON.parse(value ?? '');
  } catch (e) {
    if (!isItRetry) return parseToDataType(`"${value?.replaceAll?.('"', '')}"`, true);
    return undefined;
  }
};

export const parseToCookieType = (value) => {
  if(typeof value === 'string') return value;

  return JSON.stringify(value);
}

export const getCookie = (name) => {
  const value = `; ${document.cookie}`;

  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType(cookie.split(';')[0]) : cookie;
};

export const getCookies = (cookies = []) => {
  if (cookies.length) return cookies.reduce((result, cookie) => ({ ...result, [cookie]: getCookie(cookie) }), {});

  return Object.fromEntries(document.cookie.split('; ').map((c) => {
    const [key, value] = c.split('=');
    return [key, parseToDataType(value)]
  }));
};

export const setCookie = (name, value, expireDays) => {
  const date = new Date();

  date.setTime(date.getTime() + expireDays * 24 * 60 * 60 * 1000);
  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name) => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
