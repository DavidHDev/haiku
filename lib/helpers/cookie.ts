export const parseToDataType = <T>(
  value: string | undefined,
  isItRetry = false,
): T | undefined => {
  try {
    return value === 'undefined' || value === undefined
      ? undefined
      : (JSON.parse(value) as T);
  } catch (e) {
    if (!isItRetry) {
      return parseToDataType<T>(`"${value?.replaceAll?.('"', '')}"`, true);
    }
    return undefined;
  }
};

export const parseToCookieType = <T>(value: T): string => {
  if (typeof value === 'string') return value;
  return JSON.stringify(value);
};

export const getCookie = <T>(name: string): T | undefined => {
  const value = `; ${document.cookie}`;
  const [, cookie] = value.split(`; ${name}=`);

  return cookie ? parseToDataType<T>(cookie.split(';')[0]) : undefined;
};

export const getCookies = <T>(
  cookies: string[] = [],
): Record<string, T | undefined> => {
  if (cookies.length) {
    return cookies.reduce(
      (result, cookie) => ({ ...result, [cookie]: getCookie<T>(cookie) }),
      {} as Record<string, T | undefined>,
    );
  }

  return Object.fromEntries(
    document.cookie.split('; ').map((c) => {
      const [key, value] = c.split('=');

      return [key, parseToDataType<T>(value)];
    }),
  ) as Record<string, T | undefined>;
};

export const setCookie = <T>(
  name: string,
  value: T,
  expireDays: number,
): void => {
  const date = new Date();
  const millisecondsInADay = 24 * 60 * 60 * 1000;
  const expireDate = date.getTime() + expireDays * millisecondsInADay;

  date.setTime(expireDate);

  const expires = `expires=${date.toUTCString()}`;

  document.cookie = `${name}=${parseToCookieType(value)}; ${expires}; path=/;`;
};

export const deleteCookie = (name: string): void => {
  document.cookie = `${name}=; path=/; Expires=Thu, 01 Jan 1970 00:00:01 GMT;`;
};
