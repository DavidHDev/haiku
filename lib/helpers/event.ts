export const on = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.addEventListener)
    target.addEventListener(type, callback, options);
};

export const off = <T extends EventTarget>(
  target: T,
  type: string,
  callback: EventListenerOrEventListenerObject | null,
  options?: AddEventListenerOptions | boolean
) => {
  if (target && target.removeEventListener)
    target.removeEventListener(type, callback, options);
};
