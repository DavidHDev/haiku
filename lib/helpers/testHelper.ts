import { ReactNode } from "react";
import { render as rtlRender } from "@testing-library/react";
import { expect } from "vitest";

export const takeSnapshot = (ui: ReactNode) => {
  const { asFragment } = rtlRender(ui);
  expect(asFragment()).toMatchSnapshot();
};
