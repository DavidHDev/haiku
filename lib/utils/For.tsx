import { Children, type ReactNode } from 'react';

type RenderFn<T> = (item: T, index?: number) => ReactNode;

type Props<T> = {
  render: RenderFn<T>;
  each: T[];
};

export const For = <T,>({ render, each }: Props<T>) =>
  Children.toArray(each.map((item, index) => render(item, index)));
