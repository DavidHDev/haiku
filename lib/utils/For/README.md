`<For />`

    - Added Typescript support
    - Unit Tested using Jest + React Testing Library

Adding typescript support helps us with _auto-completion_ very-well:

![Auto Completion Example](https://i.ibb.co/jGTmLPx/Screenshot-41.png)
