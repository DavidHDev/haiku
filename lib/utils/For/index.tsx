import { Children } from 'react'

interface IProps<T> {
    each: T[]
    render: (data: T, index: number) => React.ReactNode
}

export function For<T>({ each, render }: IProps<T>): JSX.Element {
    // @todo `key={}`

    return (
        <>{Children.toArray(each.map((item, index) => render(item, index)))}</>
    )
}
