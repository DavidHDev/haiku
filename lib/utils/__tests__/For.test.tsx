import { describe, expect, it } from 'vitest';
import { takeSnapshot } from 'lib/helpers/testHelper';
import { For } from '../For';
import { render, screen } from '@testing-library/react';

describe('For', () => {
  const testArray = ['foo', 'bar', 'baz'];

  it('renders the For component', () => {
    takeSnapshot(
      <For
        each={testArray}
        render={(item, index) => <span key={index}>{`${index}: ${item}`}</span>}
      />,
    );

    testArray.forEach((item, index) => {
      expect(screen.getByText(`${index}: ${item}`)).toBeInTheDocument();
    });
    expect(document.body.firstChild!.childNodes).toHaveLength(testArray.length);
  });

  it('should not render anything if we pass empty array as prop', () => {
    render(
      <For
        each={[]}
        render={(item, index) => <span key={index}>{item}</span>}
      />,
    );
    expect(document.body.firstChild!.childNodes).toHaveLength(0);
  });
});
