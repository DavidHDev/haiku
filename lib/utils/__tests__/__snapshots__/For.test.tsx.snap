// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`For > renders the For component 1`] = `
<DocumentFragment>
  <span>
    0: foo
  </span>
  <span>
    1: bar
  </span>
  <span>
    2: baz
  </span>
</DocumentFragment>
`;
